pulp_python/app/pypi/views.py

```diff
@@ -0,0 +1,267 @@
+    SummarySerializer,
+    PackageMetadataSerializer,
+    PackageUploadSerializer,
+    PackageUploadTaskSerializer
+)
+from pulp_python.app.utils import (
+    write_simple_index,
+    write_simple_detail,
+    python_content_to_json,
+    PYPI_LAST_SERIAL,
+    PYPI_SERIAL_CONSTANT,
+)
+
+from pulp_python.app import tasks
+
+log = logging.getLogger(__name__)
+
+BASE_CONTENT_URL = urljoin(settings.CONTENT_ORIGIN, settings.CONTENT_PATH_PREFIX)
+
+
+class PyPIMixin:
+    """Mixin to get index specific info."""
+
+    @staticmethod
+    def get_repository_version(distribution):
+        """Finds the repository version this distribution is serving."""
+        pub = distribution.publication
+        rep = distribution.repository
+        if pub:
+            return pub.repository_version or pub.repository.latest_version()
+        elif rep:
+            return rep.latest_version()
+        else:
+            raise Http404("No repository associated with this index")
+
+    @staticmethod
+    def get_distribution(path):
+        """Finds the distribution associated with this base_path."""
+        distro_qs = PythonDistribution.objects.select_related(
+            "repository", "publication", "publication__repository_version"
+        )
+        try:
+            return distro_qs.get(base_path=path)
+        except ObjectDoesNotExist:
+            raise Http404(f"No PythonDistribution found for base_path {path}")
+
+    @staticmethod
+    def get_content(repository_version):
+        """Returns queryset of the content in this repository version."""
+        return PythonPackageContent.objects.filter(pk__in=repository_version.content)
+
+    def should_redirect(self, distro, repo_version=None):
+        """Checks if there is a publication the content app can serve."""
+        if distro.publication:
+            return True
+        rv = repo_version or self.get_repository_version(distro)
+        return PythonPublication.objects.filter(repository_version=rv).exists()
+
+    def get_drvc(self, path):
+        """Takes the base_path and returns the distribution, repository_version and content."""
+        distro = self.get_distribution(path)
+        repo_ver = self.get_repository_version(distro)
+        content = self.get_content(repo_ver)
+        return distro, repo_ver, content
+
+
+class PackageUploadMixin(PyPIMixin):
+    """A Mixin to provide package upload support."""
+
+    def upload(self, request, path):
+        """
+        Upload a package to the index.
+
+        0. Check request is in correct format
+        1. Check if the index allows uploaded packages (live-api-enabled)
+        2. Check if the package is in the repository already
+        3. If present then reject request
+        4. Spawn task to add content if no/old session present
+        5. Add uploads to current session to group into one task
+        """
+        live_enabled = True
+        distro = self.get_distribution(path)
+        if live_enabled:
```

Comment:
If you just invert this check, you could dedent this big block: less nesting improves the readability
```suggestion
        if not live_enabled:
            return HttpResponseForbidden(reason="Index is not editable")
```

Reply:
Good suggestion, also reminded me that I forgot to switch that conditional to the distribution's `allow_uploads` field from the hard-coded value.